limits/filetransfer/limits.go

```diff
@@ -0,0 +1,391 @@
+		return ErrFileOwnerMismatch
+	}
+
+	if !utils.FileGroupMatches(filePath, p.limits.FileTransfer.GroupGet) {
+		return ErrFileGroupMismatch
+	}
+
+	if !p.limits.FileTransfer.FollowSymLinks {
+		absolutePath, err := filepath.EvalSymlinks(filePath)
+		if err != nil {
+			return err
+		} else {
+			if absolutePath != filePath {
+				return ErrFollowLinksForbidden
+			}
+		}
+	}
+
+	if p.limits.FileTransfer.MaxFileSize > 0 {
+		fileSize := utils.FileSize(filePath)
+		if fileSize > 0 && p.limits.FileTransfer.MaxFileSize < uint64(fileSize) {
+			return ErrFileTooBig
+		}
+	}
+
+	return nil
+}
+
+func (p *Permit) BytesSent(n uint64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	countersMutex.Lock()
+	defer countersMutex.Unlock()
+
+	belowLimit = true
+	if n != 0 {
+		if deviceCountersLastH.bytesTransferred < math.MaxUint64-n {
+			deviceCountersLastH.bytesTransferred += n
+		}
+	}
+	if p.limits.FileTransfer.Counters.MaxBytesTxPerHour > 0 &&
+		deviceCountersLastH.bytesTransferred >= p.limits.FileTransfer.Counters.MaxBytesTxPerHour {
+		belowLimit = false
+	}
+
+	p.countersMutex.Lock()
+	defer p.countersMutex.Unlock()
+	if n != 0 {
+		if p.counters.bytesTransferred < math.MaxUint64-n {
+			p.counters.bytesTransferred += n
+		}
+	}
+	return belowLimit
+}
+
+func (p *Permit) BytesReceived(n uint64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	countersMutex.Lock()
+	defer countersMutex.Unlock()
+
+	belowLimit = true
+	if n != 0 {
+		if deviceCountersLastH.bytesReceived < math.MaxUint64-n {
+			deviceCountersLastH.bytesReceived += n
+		}
+	}
+	if p.limits.FileTransfer.Counters.MaxBytesRxPerHour > 0 &&
+		deviceCountersLastH.bytesReceived >= p.limits.FileTransfer.Counters.MaxBytesRxPerHour {
+		belowLimit = false
+	}
+
+	p.countersMutex.Lock()
+	defer p.countersMutex.Unlock()
+	if n != 0 {
+		if p.counters.bytesReceived < math.MaxUint64-n {
+			p.counters.bytesReceived += n
+		}
+	}
+	return belowLimit
+}
+
+func (p *Permit) BelowMaxAllowedFileSize(offset int64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	if offset < 0 {
+		return true
+	}
+	if p.limits.FileTransfer.MaxFileSize > 0 &&
+		uint64(offset) >= p.limits.FileTransfer.MaxFileSize {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (p *Permit) PreserveModes(path string, mode os.FileMode) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	if (mode & S_ISUID) != 0 {
+		mode &= os.ModePerm
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
+		}
+		mode |= os.ModeSetuid
+	} else {
+		if p.limits.FileTransfer.Umask != "" {
+			umask, err := strconv.ParseUint(p.limits.FileTransfer.Umask, 8, 32)
+			if err != nil {
+				return err
+			}
+
+			mode = os.ModePerm ^ os.FileMode(uint32(os.ModePerm)&uint32(umask))
```

Comment:
Is umask supposed to override the parameters?

Reply:
open to suggestions what do you think? my reasoning when doing that was: when I set umask in the config file limits I do it to override what comes from upstream, as in: the security policy on the device should take precedence.

Reply:
Sorry, this was a misunderstanding on my part. I forgot that this was under the context of "PreserveModes"...